solver/src/solver/paraswap_solver.rs

```diff
@@ -57,40 +60,95 @@ impl ParaswapSolver {
     }
 }
 
+pub struct ParaswapCompleteResponse {
+    transaction_result: Result<TransactionBuilderResponse, ParaswapResponseError>,
+    price_response: PriceResponse,
+    amount: U256,
+}
+
 #[async_trait::async_trait]
 impl SingleOrderSolving for ParaswapSolver {
     async fn settle_order(&self, order: LimitOrder) -> Result<Option<Settlement>> {
+        let transaction: TransactionBuilderResponse;
+        let price_response: PriceResponse;
+        let amount: U256;
+        let mut num_retries = 0;
+        loop {
```

Comment:
I'm not a huge fan of this unbound loop, and it could avoided by moving this bit to a helper function and use a `for _ in 0..MAX_RETRIES`.

Reply:
as much as I like the suggestion to use a for loop here, I am not sure if can be used in combination with @vkgnosis's good suggestion of 

```
let (x, y, z) = loop {
  if condition {
    break (1, 2, 3);
  }
}

I feel like the for loop is a bit strange for returning an Error if the iteration does indeed reach the end. Any strong preference?

Reply:
However, now that I plan to include a few different cases where we retry, this loop and match statement is becoming a bit redundant

```rust
                Err(ParaswapResponseError::PriceChange) => {
                    if num_retries >= 2 {
                        return Err(anyhow!("Price change error after {} retries", num_retries));
                    }
                    num_retries += 1;
                    continue;
                }
                Err(ParaswapResponseError::BuildingTransaction(query)) => {
                    if num_retries >= 2 {
                        return Err(anyhow!("Error Building transaction after {} retries: {}", num_retries, query));
                    }
                    num_retries += 1;
                    continue;
                }
```

Perhaps it is best to go with the for loop inside a helper method.

Reply:
The small problem with offloading this portion into a helper method is the subtle optimization performed between.

```
            let complete_response = self.paraswap_transaction_from_order(&order).await?;
            if !satisfies_limit_price(&order, &complete_response.price_response) {
                tracing::debug!("Order limit price not respected");
                return Ok(None);
            }
            match complete_response.transaction_result {
```

We have this limit price check that returns early with Ok(None) and this wont match the type we expect from the `build_or_retry()` method. 

We are already losing some of the optimization we had before, where we would return None settlement if the price query were off but now do the price and transaction query before checking the limit price. If we do this again (i.e. move the limit price check even further out) we will be making at most `max_retries` price and transaction requests before the limit price check and thus, lose all possibility of an early return.

This is not a huge issue as of now, but I would just like to point out what we lose from going this route.

Reply:
Ok, my most recent commit involving the helper method with for loop has indeed made things much better and overall more robust! Thanks for the suggestion.